src/main/java/org/truffleruby/core/binding/BindingNodes.java

```diff
@@ -414,4 +414,16 @@ protected RubyBinding binding(VirtualFrame frame) {
             return BindingNodes.createBinding(getContext(), callerFrame);
         }
     }
+
+    @Primitive(name = "caller_binding_with_new_frame")
+    public abstract static class CallerBindingWithNewFrameNode extends PrimitiveArrayArgumentsNode {
+
+        @Child ReadCallerFrameNode callerFrameNode = new ReadCallerFrameNode();
+
+        @Specialization
+        protected RubyBinding binding(VirtualFrame frame) {
+            final MaterializedFrame callerFrame = callerFrameNode.execute(frame);
+            return BindingNodes.createBinding(getContext(), newFrame(callerFrame));
```

Comment:
This creates a new Frame, and `eval()` too, shouldn't it be just one of them?

Reply:
yeah, this a bit confusing:

We have two options: `eval(source)` and `eval(source, b)`. Let's talk about refinements and forget about local vars (because the extra frame here does not affect them):

### eval(source, b)

For explicit binding on `EvalNode` level we have:
```java
// original frame is the binding frame. we do NOT create any extra frames:
final MaterializedFrame originalFrame = binding.getFrame();

// parentFrame is a frame based on binding#frame
final MaterializedFrame parentFrame = BindingNodes.newFrame(binding, newBindingDescriptor);
final Object result = eval(..);

// here we synchronize originalFrame and parentFrame changed by eval; 
// parentFrame may has activated refinements and we copy them to the originalFrame
// so bindint#frame will have these activated for parentFrame refinements;
// 
// Example:
// b = binding
// puts eval(%{using M; "hello world".upcase}, b)
// => "dlrow olleh"
// puts "hello world".upcase
// => "dlrow olleh'
syncOriginalFrame(originalFrame, parentFrame);
```


### eval(source)

For implicit binding on `EvalNode` level we have:
```java
// original frame is NOT actually binding frame, but `newFrame` created by Primitive#caller_binding_with_new_frame
final MaterializedFrame originalFrame = binding.getFrame();

// parentFrame is a frame based on binding#frame
final MaterializedFrame parentFrame = BindingNodes.newFrame(binding, newBindingDescriptor);
final Object result = eval(..);

// here we synchronize originalFrame and parentFrame changed by eval; 
// however, since parentFrame is like newFrame(binding.getFrame);
// we will NOT activate refinements for the caller binding
// 
// Example:
// b = binding
// puts eval(%{using M; "hello world".upcase})
// => "dlrow olleh"
// puts "hello world".upcase
// => "Hello world"
syncOriginalFrame(originalFrame, parentFrame);
```

As I said in the of [the comment](https://github.com/oracle/truffleruby/pull/2117#issuecomment-706667569) , I'm not sure about footprint here (becase we always create a newFrame in `#caller_binding_with_new_frame`